### ISSUE TITLE
Citations are not linked in the graph view

### ISSUE THREAD

**Issue by goktug97** [jethrokuan/org-roam#540]
I am building a web application for graph view and will publish it to Github when it is somewhat ready.
![image](https://user-images.githubusercontent.com/20567087/80625742-4fde2300-8a56-11ea-977d-cb86f3d8f7b0.png)

 I changed the `org-roam-graph--dot` function to build a JSON formatted graph. Before starting to build this application I've realized that the citations are not linked in the graph view. The citations are not linked in my application too as I am using the same function.
Citations are inserted using helm-bibtex function and I create notes for them using org-roam-bibtex

I put a `(print edge)` after the `dolist` function and it prints nothing. I also printed the `edges-cites` and it is `nil`
https://github.com/jethrokuan/org-roam/blob/bd4b9d41e8c3c5bbbbacbeb4d8dae6d0bf3278e3/org-roam-graph.el#L191-L192

I commented the `(= links:to refs:ref)` part and now it is somewhat working. But there is one edge that shouldn't be exist. I've checked both files and they don't link to each other. I've rebuilt the database and it is still the same. 
```elisp
           (edges-cites-query
            `[:with selected :as [:select [file] :from ,node-query]
                    :select :distinct [file from] :from links
                    :inner :join refs :on (and ;(= links:to refs:ref)
                                               (= links:type "cite"))
                    :where (and (in file selected) (in from selected))])

```

**Comment by goktug97** [comment]
Published it to Github
https://github.com/goktug97/org-roam-graph-server

**Comment by progfolio** [comment]
Related: #494

**Comment by myshevchuk** [comment]
Cool!

**Comment by myshevchuk** [comment]
It seems `dot` of graphviz can output directly into JSON.

https://graphviz.gitlab.io/_pages/doc/info/output.html

**Comment by TimQuelch** [comment]
That `(= links:to refs:ref)` predicate is the one that actually joins the `refs` and `links` table correctly. This is required otherwise it basically matches all your citations against all your refs. This would be why you are getting an extra link that shouldn't be there.

**Comment by goktug97** [comment]
> That `(= links:to refs:ref)` predicate is the one that actually joins the `refs` and `links` table correctly. This is required otherwise it basically matches all your citations against all your refs. This would be why you are getting an extra link that shouldn't be there.

That explains a lot but with that line `edges-cites` is empty. When I open a citation's note in emacs I can see where it is referenced in the org-roam buffer but they are not linked in the graph.

**Comment by goktug97** [comment]
> It seems `dot` of graphviz can output directly into JSON.
> 
> [graphviz.gitlab.io/_pages/doc/info/output.html](https://graphviz.gitlab.io/_pages/doc/info/output.html)
> 
> So you perhaps can use the existing `org-roam` infrastructure, at most adapt `org-roam-graph--build` to your task.

The JSON generated from the dot file would be different than I need so
I changed `org-roam-graph--dot` to build the required JSON format.

**Comment by TimQuelch** [comment]
> That explains a lot but with that line `edges-cites` is empty. When I open a citation's note in emacs I can see where it is referenced in the org-roam buffer but they are not linked in the graph.

I think I've figured it out. When a file has `#+ROAM_KEY: cite:testkey` it is stored in the `refs` table with the `ref` of `cite:testkey`. However when I cite this key from another file with `cite:testkey`, it is stored in the `links` table with a `to` of `testkey`. Therefore the join is not matching these up.

IMO I think the `cite:` prefix (or rather, any org-ref citation prefix) should be stripped from the `ROAM_KEY` before storing it in the DB

**Comment by jethrokuan** [comment]
> IMO I think the cite: prefix (or rather, any org-ref citation prefix) should be stripped from the ROAM_KEY before storing it in the DB

we can't do this now, because the refs table has no notion of type for the `ROAM_KEY`. if we strip `cite:` from the refs table, then we won't know if it's a citation or not.

**Comment by TimQuelch** [comment]
Fair enough. Then I guess the problem needs to be solved in the SQL join.

Do you know if you can join on something like the pseudocode `links.to (is suffix of refs.ref)`?

**Comment by jethrokuan** [comment]
The solution is probably to add a column to the refs table instead, which is fine.

**Comment by goktug97** [comment]
https://github.com/goktug97/org-roam/commit/7797ddea5cf89c30226a384ffb13eba7d7bf9a6d
I added key as a column it works but now I get an unwanted node in the graph. 
![image](https://user-images.githubusercontent.com/20567087/80795140-54b0ed00-8ba4-11ea-9039-18f1c217a924.png)

**Comment by goktug97** [comment]
https://github.com/goktug97/org-roam/blob/8e6cafa3b2e9cbf79ad67643ba137f9ed26a4202/org-roam-db.el#L253
https://github.com/goktug97/org-roam/blob/8e6cafa3b2e9cbf79ad67643ba137f9ed26a4202/org-roam-db.el#L274
Okay, if I don't change these two functions it works correctly.
https://github.com/goktug97/org-roam/commit/8e6cafa3b2e9cbf79ad67643ba137f9ed26a4202

**Comment by goktug97** [comment]
I've just tried with the latest commit to the repo
https://github.com/jethrokuan/org-roam/commit/1ad3539c542afd715e0e6602046581c238d06269

Citations are not linked and I have that unwanted node.
![image](https://user-images.githubusercontent.com/20567087/80799044-7adb8a80-8bae-11ea-9dd3-a8fb4fadb252.png)

**Comment by zaeph** [comment]
> I've just tried with the latest commit to the repo
> [1ad3539](https://github.com/jethrokuan/org-roam/commit/1ad3539c542afd715e0e6602046581c238d06269)
> 
> Citations are not linked and I have that unwanted node.
> ![image](https://user-images.githubusercontent.com/20567087/80799044-7adb8a80-8bae-11ea-9dd3-a8fb4fadb252.png)

Oh no. 😱

Did I break anything?  Was it working before?

**Comment by goktug97** [comment]
I tried with 2.0.0 from MELPA. It doesn't have the unwanted node.
Then I tried with the current commit and unfortunately it has the unwanted node.

**Comment by goktug97** [comment]
I think I found the problem.
I have a file and in that file, I took notes about the topic that explained in the URL that appears in the unwanted node. In the file, I have a line for that URL;
`- source :: http://blog.otoro.net/2017/10/29/visual-evolution-strategies/`. So in the recent commit, something links them together. 

I looked into my agenda.org file where I have that `DONE` statement in org-roam-buffer. Its name indexed as `DONE Read [[http://blog.otoro.net/2017/10/29/visual-evolution-strategies/][Visual Evolution Strategies]`. I added `#+TITLE: Agenda` to the file and it fixed the problem.

**Comment by goktug97** [comment]
For the problem of the unlinked references, adding a column where the column is "key" if the ref is "cite:key" to the refs table, fixes the problem. Should I open it as a pull request?

**Comment by zaeph** [comment]
> For the problem of the unlinked references, adding a column where the column is "key" if the ref is "cite:key" to the refs table, fixes the problem. Should I open it as a pull request?

I think it'd be warranted, yeah.  I haven't followed this PR too closely, what do you think @jethrokuan?

**Comment by jethrokuan** [comment]
Yeah, the extra column would probably be useful elsewhere too. Might want different values in that column though to match the current parsing though, but that can come after the PR is open.

**Comment by goktug97** [comment]
FIXED #547
